### PR TITLE
[Concept Entry]: AI Stochastic Gradient Descent

### PR DESCRIPTION
**Pull request by hitesh-gitdev**
### Description

- Created concept entry for Stochastic Gradient Descent AI search algorithm.
- Closes issue - #2135 

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

File: content/ai/concepts/search-algorithms/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+1. **Shuffle Data**: Training data is randomly shuffled before each epoch to introduce randomness and prevent biases in updates.
+
+2. **Mini-Batch Selection**: A small subset (mini-batch) of data points is randomly selected from the shuffled training dataset.
+
+3. **Compute Gradient**: The gradient of the cost function is computed using the selected mini-batch of data points. The gradient represents the direction of steepest descent, guiding the algorithm towards the minimum.
+
+4. **Update Parameters**: Based on the calculated gradient, the model's parameters are updated. The learning rate determines the step size of these updates and controls the convergence speed of the algorithm.
+
+5. **Repeat**: Steps 2 to 4 are repeated for a specified number of epochs or until a convergence criterion is met.
```
Comment (SSwiniarski) [inline content/ai/concepts/search-algorithms/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:32]:
```suggestion
1. **Shuffle Data**: Training data is randomly shuffled before each epoch to introduce randomness and prevent biases in updates.
2. **Mini-Batch Selection**: A small subset (mini-batch) of data points is randomly selected from the shuffled training dataset.
3. **Compute Gradient**: The gradient of the cost function is computed using the selected mini-batch of data points. The gradient represents the direction of steepest descent, guiding the algorithm towards the minimum.
4. **Update Parameters**: Based on the calculated gradient, the model's parameters are updated. The learning rate determines the step size of these updates and controls the convergence speed of the algorithm.
5. **Repeat**: Steps 2 to 4 are repeated for a specified number of epochs or until a convergence criterion is met.
```

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@hitesh-gitdev I did a first review and left one comment.

**Review by SSwiniarski — APPROVED** [review]
@hitesh-gitdev Sending this on to a second review.

**Comment by hitesh-gitdev** [comment]
@SSwiniarski how may I expedite the second review?

File: content/ai/concepts/search-algorithms/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+- High Variance: The randomness introduced by mini-batch selection may lead to higher variance in parameter updates, resulting in noisy convergence.
+- Learning Rate Tuning: Choosing an appropriate learning rate is crucial, as a large learning rate may cause divergence, while a small one may slow down convergence.
+- Approximate Minima: SGD may not converge to the global minimum due to its stochastic updates.
+
+## Uses in Machine Learning
+
+SGD is widely used in machine learning, especially for large-scale tasks like training deep neural networks. Its efficiency in processing large datasets, coupled with the ability to escape local minima, makes it a popular choice in modern machine learning frameworks. Moreover, SGD's adaptability allows it to be used in online learning settings where new data continuously arrives.
+
+## Pseudocode for Stochastic Gradient Descent
```
Comment (KTom101) [inline content/ai/concepts/search-algorithms/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:46]: remove this header
```suggestion
```

File: content/ai/concepts/search-algorithms/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+- Convergence: SGD can escape local minima due to its stochastic nature, improving exploration of the solution space.
+- Noise Tolerance: The inherent randomness in mini-batch selection adds noise, which helps avoid convergence to undesirable solutions.
+
+## Challenges of Stochastic Gradient Descent
+
+- High Variance: The randomness introduced by mini-batch selection may lead to higher variance in parameter updates, resulting in noisy convergence.
+- Learning Rate Tuning: Choosing an appropriate learning rate is crucial, as a large learning rate may cause divergence, while a small one may slow down convergence.
+- Approximate Minima: SGD may not converge to the global minimum due to its stochastic updates.
+
+## Uses in Machine Learning
```
Comment (KTom101) [inline content/ai/concepts/search-algorithms/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:42]: Move this section (Uses in Machine Learning) to before the section on (## Advantages of Stochastic Gradient Descent)

**Review by KTom101 — CHANGES_REQUESTED** [review]
@hitesh-gitdev I've added a few changes mainly structural. Let me know if you have any questions

**Comment by hitesh-gitdev** [comment]
@SSwiniarski @KTom101 I have addressed the comments from the 2nd PR review.

**Review by KTom101 — APPROVED** [review]
@hitesh-gitdev I made a couple more minor changes directly. This looks good to merge, thank you 🎉

**Pull request merged**